sql-plugin/src/test/scala/com/nvidia/spark/rapids/HostAllocSuite.scala

```diff
@@ -673,4 +673,73 @@ class HostAllocSuite extends AnyFunSuite with BeforeAndAfterEach with
       }
     }
   }
-}
+
+  test("split should not happen immediately after fallback on memory contention") {
+
+    // It allocates a small piece of memory (1024) as a warmup, sleep 1s,
+    // then allocates a large piece of memory
+    class AllocOnAnotherThreadWithWarmup(override val thread: TaskThread,
+        override val size: Long) extends AllocOnAnotherThread(thread, size) {
+
+      override def doAlloc(): Void = {
+        RmmRapidsRetryIterator.withRetryNoSplit {
+
+          // read shuffle data from remote and put it into Spillalble Framework
+          val w = SpillableHostBuffer.apply(HostAlloc.alloc(1024, preferPinned),
+            1024, SpillPriorities.ACTIVE_BATCHING_PRIORITY)
+
+          Thread.sleep(1000)
+
+          // let's say we will do sth with the shuffle data read, e.g. concat them
+          // for simplicity, we just have one SpillableHostBuffer as input, in real case there
+          // should be multiple SpillableHostBuffer as input
+          withResource(w.getHostBuffer()) { _ =>
+            // create another buffer as destination for concat
+            val tmp = HostAlloc.alloc(size, preferPinned)
+            synchronized {
+              closeOnExcept(tmp) { _ =>
+                assert(b.isEmpty)
+                b = Option(tmp)
+              }
+            }
+          }
+
+          // at some point close w
+          w.close()
+        }
+        null
+      }
+    }
+
+    PinnedMemoryPool.initialize(0)
+    HostAlloc.initialize(10 * 1024) // memory only big enough for the one concurrent thread to pass
+
+    failAfter(Span(10, Seconds)) {
+      val thread1 = new TaskThread("thread1", 1)
+      thread1.initialize()
+      val thread2 = new TaskThread("thread2", 2)
+      thread2.initialize()
+
+      try {
+        // Actually it will require 10 * 1024 memory
+        withResource(new AllocOnAnotherThreadWithWarmup(thread2, 9 * 1024)) { a =>
+
+          // Actually it will require 10 * 1024 memory
+          withResource(new AllocOnAnotherThreadWithWarmup(thread1, 9 * 1024)) { b =>
+
+            // At this point, both threads are blocked because of memory contention,
+            // we expect both thread1 and thread2 to fall back to BUFN state, and then thread1
+            // should attempt another try before doing splitting
+
+            b.waitForAlloc()
```

Comment:
nit, so without your change, this would have caused a split exception to be thrown on one of the two threads?

Reply:
yes, if without bobby's fix https://github.com/NVIDIA/spark-rapids-jni/pull/2976